packages/toolkit/src/query/react/suspense-utils.ts

```diff
@@ -0,0 +1,66 @@
+export interface Resource<Data> {
+  data?: Data | undefined
+  isLoading?: boolean
+}
+
+export interface IdleResource<Data> extends Resource<Data> {
+  isSkipped: true
+}
+
+export interface Suspendable {
+  getSuspendablePromise(): Promise<unknown> | undefined
+}
+
+export type SuspendableResource<Data> = Resource<Data> & Suspendable
+
+export type IdleSuspendableResource<Data> = IdleResource<Data> & Suspendable
+
+export type ResolvedSuspendableResource<T> = T extends SuspendableResource<
+  infer Data
+>
+  ? Omit<T, 'isLoading' | 'data'> & {
+      data: Exclude<Data, undefined>
+      isLoading: false
+    }
+  : never
+
+export type UseSuspendAllOutput<Sus extends readonly unknown[]> = {
+  [K in keyof Sus]: Sus[K] extends IdleSuspendableResource<any>
+    ? Sus[K]
+    : Sus[K] extends SuspendableResource<any>
+    ? ResolvedSuspendableResource<Sus[K]>
+    : Sus[K] extends Suspendable
+    ? Sus[K]
+    : never
+}
+
+function isPromiseLike(val: unknown): val is PromiseLike<unknown> {
+  return (
+    !!val && typeof val === 'object' && typeof (val as any).then === 'function'
+  )
+}
+
+function getSuspendable(suspendable: Suspendable) {
+  return suspendable.getSuspendablePromise()
+}
+
+export function useSuspendAll<
+  G extends SuspendableResource<any>,
+  T extends SuspendableResource<any>[]
+>(
+  ...suspendables: readonly [G, ...T]
+): UseSuspendAllOutput<readonly [G, ...T]> {
+  if (!suspendables.length) {
+    throw new TypeError('useSuspendAll: requires one or more arguments')
+  }
+
+  let promises = suspendables
+    .map(getSuspendable)
+    .filter(isPromiseLike) as Promise<unknown>[]
+
+  if (promises.length) {
+    throw Promise.all(promises)
```

Comment:
That would create a new promise each time, but we need to keep it stable. Maybe just throw the first unresolved promise? (We would need to add an `isResolved` property to the promises in that case)

Reply:
> That would create a new promise each time, but we need to keep it stable. Maybe just throw the first unresolved promise? (We would need to add an `isResolved` property to the promises in that case)

I've not seen suspense for data fetching implementation that relies on `isResolved` or similar.

> That would create a new promise each time

Yes but it is not a problem because you are suspending the rendering until all the promises are resolved and 
`runningQueries` does not keep promises that are settled.

Moreover it is the same approach used by [joitai waitForAll](https://github.com/pmndrs/jotai/blob/main/src/utils/waitForAll.ts)

Reply:
I remember reading in the working group that for suspense it is important to always get the same promise back, but right now I can't find the quote for it.

Reply:
Yep, pretty sure I saw that too, but don't immediately have a pointer either.

Reply:
> Yep, pretty sure I saw that too, but don't immediately have a pointer either.

I'll look into it, I've got a couple of ideas on how to tackle this issue but I'd like to migrate the examples to react 18 first.